sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -627,12 +758,32 @@ sycl::detail::pi::PiProgram ProgramManager::getBuiltPIProgram(
   // getOrBuild is not supposed to return nullptr
   assert(BuildResult != nullptr && "Invalid build result");
 
+  sycl::detail::pi::PiProgram ResProgram = BuildResult->Val;
+  auto Plugin = ContextImpl->getPlugin();
+  auto UpdateRefCountF = [ResProgram, &Plugin]() {
+    // For every cached copy of the program, we need to increment its refcount
+    Plugin->call<PiApiKind::piProgramRetain>(ResProgram);
+    return ResProgram;
+  };
+  // If we linked any extra device images for virtual functions, then we need to
+  // cache them as well.
+  for (const RTDeviceBinaryImage *BImg : DeviceImagesToLink) {
+    // CacheKey is captured by reference by GetCachedBuildF, so we can simply
+    // update it here and re-use that lambda.
+    CacheKey.first.second = BImg->getImageID();
+    // TODO: Use SYCL 2020 exceptions
+    auto TempRes = Cache.getOrBuild<compile_program_error>(GetCachedBuildF,
+                                                           UpdateRefCountF);
```

Comment:
I like the use of the retain lambda here to duplicate cache entries, but I think seeing the name "getOrBuild" here is a little disorienting. Not sure what would be a better name to generalize it for this case though, may be even just changing it to "getOrCreate" would be a little better?

Reply:
> I like the use of the retain lambda here to duplicate cache entries, but I think seeing the name "getOrBuild" here is a little disorienting. Not sure what would be a better name to generalize it for this case though, may be even just changing it to "getOrCreate" would be a little better? Or we can consider this a corner use case and just leave a clarifying comment here.

This code is copied from `kernel_bundle::build` implementation for multiple devices and I agree that the naming is bad, because I spent a couple of days debugging why caching doesn't work and it turned out that I'm calling a wrong method. I think since this is a second use of such functionality, it may be better to have a dedicated method for such use case.

Reply:
I've introduce the new API in 4471e5b5611e10f6ac95ba3d304c5b829cb89e24 - could you please provide your feedback about it and once we think it is good, I can update `kernel_bundle::build` implementation to use it as well (can be done in  a separate PR, if preferred).

Reply:
That LGTM, a separate PR for that is fine. Thanks!